core/clients/key_flow_continuous_refresh_test.go

```diff
@@ -0,0 +1,354 @@
+			contextClosesIn: 100 * time.Millisecond,
+			doError: &oapierror.GenericOpenAPIError{
+				StatusCode: http.StatusInternalServerError,
+			},
+			expectedNumberDoCalls: 2,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			numberDoCalls := 0
+			mockDo := func(client *http.Client, req *http.Request, cfg *RetryConfig) (resp *http.Response, err error) {
+				numberDoCalls++
+
+				if tt.doError != nil {
+					return nil, tt.doError
+				}
+
+				accessToken, err := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.RegisteredClaims{
+					ExpiresAt: jwt.NewNumericDate(time.Now().Add(accessTokensTimeToLive)),
+				}).SignedString([]byte("test"))
+				if err != nil {
+					t.Fatalf("Do call: failed to create access token: %v", err)
+				}
+
+				responseBodyStruct := TokenResponseBody{
+					AccessToken: accessToken,
+				}
+				responseBody, err := json.Marshal(responseBodyStruct)
+				if err != nil {
+					t.Fatalf("Do call: failed to marshal response: %v", err)
+				}
+				response := &http.Response{
+					StatusCode: http.StatusOK,
+					Body:       io.NopCloser(bytes.NewReader(responseBody)),
+				}
+				return response, nil
+			}
+
+			accessToken, err := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.RegisteredClaims{
+				ExpiresAt: jwt.NewNumericDate(time.Now().Add(accessTokensTimeToLive)),
+			}).SignedString([]byte("test"))
+			if err != nil {
+				t.Fatalf("failed to create access token: %v", err)
+			}
+
+			ctx := context.Background()
+			ctx, cancel := context.WithTimeout(ctx, tt.contextClosesIn)
+			defer cancel()
+
+			keyFlow := &KeyFlow{
+				config: &KeyFlowConfig{
+					ClientRetry:                   NewRetryConfig(),
+					BackgroundTokenRefreshContext: ctx,
+				},
+				doer: mockDo,
+				token: &TokenResponseBody{
+					AccessToken: accessToken,
+				},
+			}
+
+			refresher := &continuousTokenRefresher{
+				keyFlow:                        keyFlow,
+				timeStartBeforeTokenExpiration: timeStartBeforeTokenExpiration,
+				timeBetweenContextCheck:        timeBetweenContextCheck,
+				timeBetweenTries:               timeBetweenTries,
+			}
+
+			err = refresher.continuousRefreshToken()
+			if err == nil {
+				t.Fatalf("routine finished with non-nil error")
+			}
+			if numberDoCalls != tt.expectedNumberDoCalls {
+				t.Fatalf("expected %d calls to API to refresh token, got %d", tt.expectedNumberDoCalls, numberDoCalls)
+			}
+		})
+	}
+}
+
+// Tests if
+//   - continuousRefreshToken() changes the token
+//   - The access token can be accessed while continuousRefreshToken() is trying to update it
+func TestContinuousRefreshTokenConcurrency(t *testing.T) {
+	// The times here are in the order of miliseconds (so they run faster)
+	// For this to work, we need to increase precision of the expiration timestamps
+	jwt.TimePrecision = time.Millisecond
+
+	// Test plan:
+	// 1) refreshToken() will trigger a token update. It will be blocked in the mockDo() routine
```

Comment:
why do we need to artificially block the token update? Wouldn't it be possible to write the test in a way that the refreshToken() routine is waiting for the refreshTimestamp "naturally" while the first request is done?

Maybe Im missing something

Reply:
It's intended, to test that the access token is still usable even when `continuousRefreshToken()` is trying to update the token